test/test_testing.py

```diff
@@ -2706,3 +2717,465 @@ def test_stop_stopped_service(self):
+        with self.assertRaises(pebble.PathError) as cm:
+            client.make_dir(self.prefix + '/file/subdir/subdir', make_parents=True)
+        self.assertEqual(cm.exception.kind, 'generic-file-error')
+
+    def test_make_dir_with_permission_mask(self):
+        client = self.client
+        client.make_dir(self.prefix + '/dir1', permissions=0o700)
+        client.make_dir(self.prefix + '/dir2', permissions=0o777)
+
+        files = client.list_files(self.prefix + '/', pattern='dir*')
+        self.assertEqual([f for f in files if f.path == self.prefix + '/dir1']
+                         [0].permissions, 0o700)
+        self.assertEqual([f for f in files if f.path == self.prefix + '/dir2']
+                         [0].permissions, 0o777)
+
+        # If permissions are outside of the range 0o000 through 0o777, an exception should be
+        # raised.
+        for i, bad_permission in enumerate((
+            0o1000,  # Exceeds 0o777
+            -1,      # Less than 0o000
+        )):
+            with self.assertRaises(pebble.PathError) as cm:
+                client.make_dir(self.prefix + '/dir3_{}'.format(i), permissions=bad_permission)
+            self.assertEqual(cm.exception.kind, 'generic-file-error')
+
+    def test_remove_path(self):
+        client = self.client
+        client.push(self.prefix + '/file', '')
+        client.make_dir(self.prefix + '/dir/subdir', make_parents=True)
+        client.push(self.prefix + '/dir/subdir/file1', '')
+        client.push(self.prefix + '/dir/subdir/file2', '')
+        client.push(self.prefix + '/dir/subdir/file3', '')
+        client.make_dir(self.prefix + '/empty_dir')
+
+        client.remove_path(self.prefix + '/file')
+
+        client.remove_path(self.prefix + '/empty_dir')
+
+        # Remove non-empty directory, recursive=False: error
+        with self.assertRaises(pebble.PathError) as cm:
+            client.remove_path(self.prefix + '/dir', recursive=False)
+        self.assertEqual(cm.exception.kind, 'generic-file-error')
+
+        # Remove non-empty directory, recursive=True: succeeds (and removes child objects)
+        client.remove_path(self.prefix + '/dir', recursive=True)
+
+        # Deliberately ignoring a few cases right now, as the behavior for these may
+        # change based upon discussions:
+        # * Removing non-existent path, recursive=False: currently does error
+        # * Removing non-existent path, recursive=True: currently does not error
+
+    # Other notes:
+    # * Parent directories created via push(make_dirs=True) default to root:root ownership
+    #   and whatever permissions are specified via the permissions argument; as we default to None
+    #   for ownership/permissions, we do ignore this nuance.
+    # * Parent directories created via make_dir(make_parents=True) default to root:root ownership
+    #   and 0o755 permissions; as we default to None for ownership/permissions, we do ignore this
+    #   nuance.
```

Comment:
creating directories and then populating them ignores the permissions of the directories that were created